src/server.rs

```diff
@@ -68,43 +125,43 @@ where
 
     fn call(&self, req: http::Request<A>) -> Self::Future {
         let is_gzip = is_gzip(&req);
-        Box::new(self.inner.call(req).map(move |rsp| {
+        trace!("GzipChunked: is_gzip={:?};", is_gzip);
+        let chunksz = self.chunk_size;
+        let compression = self.compression;
+
+        Box::new(self.inner.call(req).and_then(move |rsp| {
             let (mut parts, body) = rsp.into_parts();
+
             let body: Body = if is_gzip {
                 let mut encoder = GzEncoder::new(
-                    body.as_ref(),
-                    Compression::default());
-                parts.headers.insert(header::CONTENT_ENCODING, "gzip".parse().unwrap());
-                // let (mut tx_body, rx_body) = Body::channel();
-                // executor::spawn(future::lazy(move || {
-                //     let mut buf = [0u8; 4096];
-                //     loop {
-                //         match encoder.read(&mut buf) {
-                //             Ok(n) if n == 0 => { break; },
-                //             Ok(n) => {
-                //                 let chunk: hyper::Chunk = buf[0..n].to_vec().into();
-                //                 // while let Ok(Async::NotReady) = tx_body.poll_ready() { };
-                //                 tx_body.send_data(chunk)
-                //                     .expect("send data");
-                //             },
-                //             Err(e) => {
-                //                 eprintln!("error: {:?}", e);
-                //                 break;
-                //                 }
-                //         }
-                //     }
-                //     future::ok::<(),()>(())
-                // }));
-                let mut buf = Vec::new();
-                encoder.read_to_end(&mut buf);
-                buf.into()
+                    Cursor::new(Vec::<u8>::new()),
+                    compression,
+                );
+
+                let n = encoder.write(body.as_ref())
+                    .map_err(hyper::Error::Io)?;
+                let mut read = encoder.finish()
+                    .map_err(hyper::Error::Io)?;;
+                trace!("GzipChunked: encoded {:?} bytes", n);
+
+                read.set_position(0);
```

Comment:
It seems you already have the full buffer in memory, so there's not much need to use `ChunkingStream`. It will just cause a second copy, by copying from the `Cursor<Vec>`, into a `Bytes`.

Reply:
Yeah, I think I've somewhat misunderstood how to use `flate2` --- the goal is to be able to do all the compression streaming, and only buffer the compressed output up to the chunk size.

What I'd _really_ like is to be able to pass a writer to the inner service that's either the encoder or an uncompressed stream, so that I can just `write!` to it, but I haven't quite figured out a way to get the type system to let me do that yet...